python/cuml/fil/fil.pyx

```diff
@@ -0,0 +1,232 @@
+
+from libcpp cimport bool
+from libc.stdint cimport uintptr_t
+from libc.stdlib cimport calloc, malloc, free
+
+from cuml.common.base import Base
+from cuml.common.handle cimport cumlHandle
+from cuml.utils import get_cudf_column_ptr, get_dev_array_ptr, \
+    input_to_dev_array, zeros
+cimport cuml.common.handle
+cimport cuml.common.cuda
+
+cdef extern from "treelite/c_api.h":
+    ctypedef void* ModelHandle
+
+cdef extern from "fil/fil.h" namespace "ML::fil":
+
+    cdef enum algo_t:
+        NAIVE,
+        TREE_REORG,
+        BATCH_TREE_REORG
+
+    cdef enum output_t:
+        RAW,
+        PROB,
+        CLASS
+
+    cdef struct dense_node_t:
+        float val
+        int bits
+
+    cdef struct forest:
+        pass
+
+    ctypedef forest* forest_t
+
+    cdef struct treelite_params_t:
+        algo_t algo
+        bool output_class
+        float threshold
+
+    cdef void free(cumlHandle& handle,
+                   forest_t)
+
+    cdef void predict(cumlHandle& handle,
+                      forest_t,
+                      float*,
+                      float*,
+                      size_t)
+
+    cdef forest_t from_treelite(cumlHandle& handle,
+                                forest_t*,
+                                ModelHandle,
+                                treelite_params_t*)
+
+cdef class FIL_impl():
+
+    cpdef object handle
+    cdef forest_t forest_data
+    cdef object algo
+    cdef object threshold
+
+    def __cinit__(self,
+                  algo=0,
+                  threshold=0.0,
+                  handle=None):
+        self.algo = algo
+        self.threshold = threshold
+        self.handle = handle
+
+    def predict(self, X, preds=None):
+        cdef uintptr_t X_ptr
+        X_m, X_ptr, n_rows, _, X_dtype = \
+            input_to_dev_array(X, order='C')
+
+        cdef cumlHandle* handle_ =\
+            <cumlHandle*><size_t>self.handle.getHandle()
+
+        if preds is None:
+            preds = cudf.Series(zeros(n_rows, dtype=np.float32))
+        cdef uintptr_t preds_ptr
+        preds_m, preds_ptr, _, _, _ = \
+            input_to_dev_array(preds)
+        predict(handle_[0],
+                self.forest_data,
+                <float*> preds_ptr,
+                <float*> X_ptr,
+                <size_t> n_rows)
+        self.handle.sync()
+        # synchronous w/o a stream
+        return preds
+
+    def from_treelite(self, model, output_class):
+        cdef treelite_params_t treelite_params
+        treelite_params.output_class = output_class
+        treelite_params.threshold = self.threshold
+        treelite_params.algo = self.algo
+        cdef forest_t* forest_pointer =\
+            <forest_t*><size_t> model.handle.value
+        cdef cumlHandle* handle_ =\
+            <cumlHandle*><size_t>self.handle.getHandle()
+        cdef uintptr_t model_ptr = model.handle.value
+        self.forest_data = from_treelite(handle_[0],
+                                         forest_pointer,
```

Comment:
Doesn't something like `&self.forest_data` work? If it doesn't, can `forest_data` be declared a local variable and then `&forest_data`, same as `&treelite_params`?

If it works, the modification to the C++ `from_treelite()` becomes unnecessary.

Reply:
i tried to use the updated `forest_pointer` to assign value to the `self.forest_data`  but when i do so it does not read the data in right or i get cython compilation errors. If you dont want it to return anything then i can change the code to accept `forest_pointer` in place of `forest_data`. So the input for the `predict` function will accept data of type `forest_t*` in place of `forest_t`

Reply:
@canonizer Just did a quick offline review regarding this wish Saloni, correction should be coming soon indeed as you pointed out